solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -224,6 +232,11 @@ private Collector buildAndRunCollectorChain(QueryResult qr, Query query, Collect
     if (collector instanceof DelegatingCollector) {
       ((DelegatingCollector) collector).finish();
     }
+
+    if (cmd.isQueryCancellable()) {
```

Comment:
This code should go under the `finally` section to make sure we always properly handle de-registration of queries even if they throw exceptions.